prog/dftb+/lib_dftbplus/main.F90

```diff
@@ -6722,463 +6632,255 @@ subroutine getHamiltonianLandEnergyL(env, denseDesc, sccCalc, orb, &
-  subroutine getReksSccHamiltonian(H0, over, nNeighbourSK, &
-      & neighbourList, species, orb, iSparseStart, img2CentCell, &
-      & hamSp, intBlock, Lpaired, iL)
-
-    !> non-SCC hamitonian (sparse)
-    real(dp), intent(in) :: H0(:)
-
-    !> overlap (sparse)
-    real(dp), intent(in) :: over(:)
-
-    !> Number of atomic neighbours
-    integer, intent(in) :: nNeighbourSK(:)
-
-    !> list of atomic neighbours
-    type(TNeighbourList), intent(in) :: neighbourList
-
-    !> species of atoms
-    integer, intent(in) :: species(:)
-
-    !> atomic orbital information
-    type(TOrbitals), intent(in) :: orb
-
-    !> Index for atomic blocks in sparse data
-    integer, intent(in) :: iSparseStart(:,:)
-
-    !> image atoms to central cell atoms
-    integer, intent(in) :: img2CentCell(:)
-
-    !> resulting hamitonian (sparse)
-    real(dp), intent(out) :: hamSp(:,:)
-
-    !> internal block and spin resolved potential for each microstate
-    real(dp), intent(inout) :: intBlock(:,:,:,:)
-
-    !> Number of spin-paired microstates
-    integer, intent(in) :: Lpaired
-
-    !> currect index of loop L
-    integer, intent(in) :: iL
-
-    real(dp), allocatable :: tmpBlock(:,:,:,:)
-    integer :: nAtom
+    ! Calculate contribution to Hamiltonian except rangeseparated part
+    reks%intShellL(:,:,:,:) = 0.0_dp
+    reks%intBlockL(:,:,:,:,:) = 0.0_dp
+    do iL = 1, reks%Lmax
 
-    nAtom = size(orb%nOrbAtom)
+      ! reks%chargePerShellL has (qm) component
+      call getChargePerShell(reks%qOutputL(:,:,:,iL), orb, species,&
+          & reks%chargePerShellL(:,:,:,iL))
+      call resetInternalPotentials(tDualSpinOrbit, xi, orb, species, potential)
+      call addChargePotentials(env, sccCalc, reks%qOutputL(:,:,:,iL), q0, &
+          & reks%chargePerShellL(:,:,:,iL), orb, species, neighbourList, &
+          & img2CentCell, spinW, solvation, thirdOrd, potential, electrostatics, &
+          & tPoisson, tUpload, shiftPerLUp)
 
-    allocate(tmpBlock(orb%mOrb,orb%mOrb,nAtom,1))
+      ! reks%intShellL, reks%intBlockL has (qm) component
+      reks%intShellL(:,:,:,iL) = potential%intShell
+      reks%intBlockL(:,:,:,:,iL) = potential%intBlock
 
-    ! tmpBlock has (my_qm) component
-    ! qm representation is converted to my_qm representation
-    if (iL <= Lpaired) then
-      ! If iL = 1, then qm = 1u + 1d, 1u - 1d and my_qm = 1u + 1d
-      ! calculate charge part
-      tmpBlock(:,:,:,1) = intBlock(:,:,:,1)
-      hamSp(:,1) = h0(:)
-    else
-      if (mod(iL,2) == 1) then
-        ! If iL = 3, then qm = 3u + 3d, 3u - 3d and my_qm = 3u + 3d
+      ! qm representation is converted to my_qm representation
+      if (iL <= reks%Lpaired) then
+        ! If iL = 1, then qm = 1u + 1d, 1u - 1d and my_qm = 1u + 1d
         ! calculate charge part
-        tmpBlock(:,:,:,1) = intBlock(:,:,:,1)
-        hamSp(:,1) = h0(:)
+        potential%intBlock(:,:,:,1) = reks%intBlockL(:,:,:,1,iL)
+        tmpHamSp(:,1) = h0
```

Comment:
The getSccHamiltonian call below is intent(out) for tmpHamSp so will over-write the stored values.

Reply:
In commit b2264d1, the intent is now changed, so tmpHamSp is normally updated when it enters to getSccHamiltonian routine. The input value is H0 or zero with respect to current microstate index.